File: wot-dijkstra/src/main/java/org/pgpainless/wot/dijkstra/sq/CertSynopsis.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import org.pgpainless.algorithm.RevocationState;
import org.pgpainless.key.OpenPgpFingerprint;

public class CertSynopsis {

    private final OpenPgpFingerprint fingerprint;
    private final Date expirationTime;
    private final RevocationState revocationState;
    private final Map<String, RevocationState> userIds;

    /**
     * Create a new {@link CertSynopsis}.
     *
     * @param fingerprint fingerprint of the certificate
     * @param expirationTime expiration time
     * @param revocationState revocation state of the certificate
     * @param userIds set of user-ids
     */
    public CertSynopsis(OpenPgpFingerprint fingerprint,
                        Date expirationTime,
                        RevocationState revocationState,
                        Map<String, RevocationState> userIds) {
        this.fingerprint = fingerprint;
        this.expirationTime = expirationTime;
        this.revocationState = revocationState;
        this.userIds = userIds;
    }

    /**
     * Return the fingerprint of the certificate.
     *
     * @return fingerprint
     */
    public OpenPgpFingerprint getFingerprint() {
        return fingerprint;
    }

    /**
     * Get the certificates expiration time.
     *
     * @return expiration time
     */
    public Date getExpirationTime() {
        return expirationTime;
    }

    /**
     * Get the revocation status of the certificate.
     *
     * @return revocation state
     */
    public RevocationState getRevocationState() {
        return revocationState;
    }

    /**
     * Get a {@link Set} containing all user-ids of the certificate.
     *
     * @return user-ids
     */
    public Map<String, RevocationState> userIds() {
        return new HashMap<>(userIds);
    }

    @Override
    public String toString() {
        return fingerprint + (userIds.isEmpty() ? "" : "(" + userIds.keySet().iterator().next() + ")");
    }
}

File: wot-dijkstra/src/main/kotlin/org/pgpainless/wot/dijkstra/sq/CertSynopsis.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq

import java.util.*

/**
 * A [CertSynopsis] is a proxy object containing information about a certificate.
 *
 * @param fingerprint [Fingerprint] of the certificate
 * @param expirationTime optional expiration time of the certificate
 * @param revocationState [RevocationState] denoting whether the certificate is revoked or not
 * @param userIds [Map] of user-ids on the certificate, along with their revocation states
 */
data class CertSynopsis(
        val fingerprint: Fingerprint,
        val expirationTime: Date? = null,
        val revocationState: RevocationState = RevocationState.notRevoked(),
        val userIds : Map<String, RevocationState> = mapOf()) {

    constructor(fingerprint: String,
            expirationTime: Date? = null,
            revocationState: RevocationState = RevocationState.notRevoked(),
            userIds: Map<String, RevocationState> = mapOf()):
            this(Fingerprint(fingerprint), expirationTime, revocationState, userIds)

    override fun toString(): String {
        return if (userIds.isEmpty()) {
            "$fingerprint"
        } else {
            "$fingerprint (${userIds.keys.first()})"
        }
    }
}